### PR TITLE
Register and resolve modules from request scope

### LINKED ISSUE #61: Question about services lifecycles
I´m trying to use the Botwin library with EF Core. I registered the DbContext using services.AddDbContext (By default sets ServiceLifetime.Scoped) but I noticed that my application uses the same DbContext in each request. Then I noticed that the modules are created once, regardless of whether they are registered at IoC container as transient. So each dependency injected don't respect its configured Lifetime. Is this a desing decision?

Thanks in advance.

### PR DESCRIPTION
**Pull request by jchannon**
Fixes #61

### CODE REVIEW

**Comment by jchannon** [comment]
I think this is going to be as good as it gets for a first pass. As @khellang said there will be a lot of allocations per request with this though. I can't see how you can cache anything to improve perf here. @davidfowl, do you have any thoughts before this gets merged?

File: src/BotwinExtensions.cs [hunk 1]
```diff
@@ -164,13 +172,14 @@ public static void AddBotwin(this IServiceCollection services, params Assembly[]
             var modules = assemblies.SelectMany(x => x.GetTypes().Where(t => typeof(BotwinModule).IsAssignableFrom(t) && t != typeof(BotwinModule)));
             foreach (var module in modules)
             {
-                services.AddTransient(typeof(BotwinModule), module);
+                services.AddScoped(module);
```
Comment (davidfowl) [inline src/BotwinExtensions.cs]: Why not leave it as transient?

**Comment by jchannon** [comment]
Because the modules are not used once the route is registered with
Routebuilder. If you create a new guid in the ctor and return it in the
route it is always the same value. The ctor is only hit once at startup not
once per request so all dependencies currently become a singleton

**Comment by davidfowl** [comment]
Yea but you aren't fixing anything by making a scope. You're still creating the modules at app startup time. You don't need to make anything scoped here, they're probably only resolved once per request right? 

You just need to use the service provider on the HttpContext to resolve the module again inside of the route callback. That should be enough to solve your problems.

**Comment by jchannon** [comment]
@davidfowl Hmm OK, will take a look thanks. @khellang any thoughts?

**Comment by jchannon** [comment]
Removing the scope does still work with the module being resolved inside the wrapping handler

**Comment by davidfowl** [comment]
Hmm actually it'll break if you inject a DbContext (or any other scoped dependency). Go ahead and keep the scope for now. It's hacky but it solves this problem.

The biggest downside is the fact that you have to instantiate the module, run the constructor which to get the list of routes which are only used once at startup. You always create a single instance of a module that gets thrown away.

**Comment by jchannon** [comment]
@davidfowl is there any way to improve the perf doing that resolving, creating the actual handler each request?

**Comment by davidfowl** [comment]
I forgot to add that work also happens per request but by then nobody cares about the routes.

**Comment by jchannon** [comment]
Well the routes are looked up from the singleton module by the request scoped module to find the correct handler

**Comment by khellang** [comment]
> Hmm actually it'll break if you inject a DbContext (or any other scoped dependency).

Yeah, the scope has to be there in order to not leak scoped dependencies resolved from the app container...

File: src/BotwinExtensions.cs (hunk 1 above)
Comment (jchannon) [reply src/BotwinExtensions.cs]: @khellang why is the module added twice?

Comment (poke) [reply src/BotwinExtensions.cs]: @jchannon You need to add the module once as `BotwinModule`, so you can resolve all modules by requesting `IEnumerable<BotwinModule>` from DI. And then, you need to register each module as itself, so you can get the exact module later when you retrieve the module in the path handler.

Comment (jchannon) [reply src/BotwinExtensions.cs]: Seems odd 😄

Comment (poke) [reply src/BotwinExtensions.cs]: Maybe, but basically you are just assigning each module two “labels” you can refer it to. Once as the module itself, and once with it being *a* module.

**Comment by jchannon** [comment]
Thanks @poke @khellang @davidfowl 👍 ❤️ 👍

**Pull request merged**